File: src/packs/hints.ts
export interface SitePackHints {
  pageSignatures: Record<string, string[]>;
  knownSignals: string[];
  raw: Record<string, unknown>;
}

export function parseHints(input: unknown): SitePackHints {
  const raw = input && typeof input === 'object' && !Array.isArray(input) ? (input as Record<string, unknown>) : {};
  const pageSignatures =
    raw.page_signatures && typeof raw.page_signatures === 'object' && !Array.isArray(raw.page_signatures)
      ? Object.fromEntries(
          Object.entries(raw.page_signatures as Record<string, unknown>).map(([key, value]) => [
            key,
            Array.isArray(value) ? value.filter((item): item is string => typeof item === 'string') : []
          ])
        )
      : {};

  const knownSignals = [
    ...Object.keys(pageSignatures),
    ...Object.values(pageSignatures).flatMap((signals) => signals)
  ].slice(0, 16);

  return {
    pageSignatures,
    knownSignals,
    raw
  };
}

File: browser_platform/packs/hints.py
from __future__ import annotations


def parse_hints(input_data: object) -> dict:
    raw = input_data if isinstance(input_data, dict) and not isinstance(input_data, list) else {}

    raw_sigs = raw.get('page_signatures')
    if raw_sigs and isinstance(raw_sigs, dict) and not isinstance(raw_sigs, list):
        page_signatures = {
            k: [item for item in v if isinstance(item, str)] if isinstance(v, list) else []
            for k, v in raw_sigs.items()
        }
    else:
        page_signatures = {}

    known_signals = list(page_signatures.keys())
    for signals in page_signatures.values():
        known_signals.extend(signals)

    return {
        'pageSignatures': page_signatures,
        'knownSignals': known_signals,
        'raw': raw,
    }
